### PR TITLE
Post

### PR DESCRIPTION
**Pull request by house9**

### CODE REVIEW

File: post.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 In this blog post, we will explore how to use [Zod](https://zod.dev/), a declarative JavaScript validation library, for form validation in a [Next.js](https://nextjs.org/) application. We will also discuss the importance of server-side validation and how to handle validation errors returned from the server.
```
Comment (sarmstead) [inline post.md]: This is a strong transition statement, but we may want to add more about why this post is important. I suggest merging the paragraph from "The Importance of Server-Side Validation" into this intro paragraph. Something like, "Server-side validation is a critical line of defense against invalid and malicious data, and ensures data integrity and security. In this post, we will explore how to use [Zod](https://zod.dev/), a declarative JavaScript validation library, for server-side form validation in a [Next.js](https://nextjs.org/) application. We will also look into to handling validation errors returned from the server."

File: post.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+This article assumes you're acquainted with setting up and running Next.js applications.
+
+To view the code referenced in this article, checkout the [repo](https://github.com/CodingZeal/nextjs-server-action-validation).
+
+This article and accompanying example application utilize Next.js version 14.x configured with TailwindCSS and TypeScript.
+
+## The Importance of Server-Side Validation
+
+While client-side validation provides immediate feedback and improves user experience, server-side validation is a must for any application. It acts as a second line of defense against invalid or malicious data. Even if you augment your application with client-side validation, server-side validation is necessary to ensure data integrity and security.
```
Comment (sarmstead) [inline post.md]: We can remove this paragraph if merged into the introduction.

File: post.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+```ts
+import { z } from "zod";
+
+const schema = z.object({
+  name: z.string().min(1, "Name cannot be blank"),
+  email: z.string().email("Invalid email address"),
+});
+```
+
```
Comment (sarmstead) [inline post.md:29]: Should we explain to the reader what the "Name cannot be blank" string is doing, as well as the `.email` method?

File: post.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+## Handling Server-Side Validation Errors
+
+When we submit our form, the server-side action validates the form data against the Zod schema. If the validation fails, the server returns an array of errors (`ZodIssue[]`). Each error object has the following structure:
```
Comment (sarmstead) [inline post.md:32]: I'd suggest a transition sentence here. Something like, "Using Zod, let's tease out a validation flow that integrates with Next.js. When we submit our form, a Next.js server-side action validates ...."

File: post.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    return {
+      errors: validation.error.issues,
+    };
+  }
+}
+```
+
+## Contact Page
+
+In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a server component while the `Form` component is a client component. We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
```
Comment (sarmstead) [inline post.md]: I suggest adding doc links to the phrases _server component_ and _client component_.

File: post.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  );
+}
+```
+
+## Form Component
+
+In our form component, we use the `useFormState` hook to handle validation errors. We match the `path` node of each error object to the corresponding input field and extract all `message` nodes to build the error message for each field.
+
+```tsx
+// app/contact/form.tsx
```
Comment (sarmstead) [inline post.md:107]: Add a line return here

File: post.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+```
+
+Now, let's look at how we can implement this in our Next.js application. We'll start with the server action, then move on to the contact page, and finally the form component.
+
+## Server Action
+
+In our server action, we validate the form data using the Zod schema. If the validation fails, we return the array of errors. If the validation succeeds, we redirect the user to the home page.
+
+```ts
+// app/contact/server-action.ts
```
Comment (sarmstead) [inline post.md:53]: Add a line return here (after the path name)

File: post.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+import { useFormState, useFormStatus } from "react-dom";
+import type { ZodIssue } from "zod";
+
+type Props = {
+  action: (
+    _prevState: any,
+    params: FormData
+  ) => Promise<{ errors: ZodIssue[] }>;
+};
```
Comment (sarmstead) [inline post.md:119]: As a reader, I'd like to know what this type definition is doing. Can you explain?

File: post.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+const findErrors = (fieldName: string, errors: ZodIssue[]) => {
+  return errors
+    .filter((item) => {
+      return item.path.includes(fieldName);
+    })
+    .map((item) => item.message);
+};
+```
+
+In conclusion, Zod provides a powerful and flexible way to handle form validation in a Next.js application. By combining it with server-side actions and the `useFormState` hook, we can create robust forms with clear and specific error messages.
```
Comment (sarmstead) [inline post.md]: A few things here:
- I suggest adding a section title. Something like, "Wrapping it Up"
- Let's flesh out more how the "why" is accomplished at this point

File: post.md (hunk 4 above)
Comment (house9) [reply post.md:32]: Going to stick with my bland copy for this paragraph

File: post.md (hunk 8 above)
Comment (house9) [reply post.md:119]: Going to assume the reader is familiar with typescript already and do not want to tangent from the main concepts too much.

File: post.md [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # Using Zod for Form Validation with React Server Actions in Next.js
 
-In this blog post, we will explore how to use [Zod](https://zod.dev/), a declarative JavaScript validation library, for form validation in a [Next.js](https://nextjs.org/) application. We will also discuss the importance of server-side validation and how to handle validation errors returned from the server.
+Server-side validation is a critical line of defense against invalid and malicious data, and ensures data integrity and security. In this post, we will explore how to use [Zod](https://zod.dev/), a declarative JavaScript validation library, for server-side form validation in a [Next.js](https://nextjs.org/) application. We will also look into to handling validation errors returned from the server.
```
Comment (sarmstead) [inline post.md:3]: This is a fabulous introduction 👏🏽

File: post.md [hunk 11]
```diff
@@ -10,13 +10,11 @@ To view the code referenced in this article, checkout the [repo](https://github.
...
-
 ## Overview of Zod Schema for Validation
 
-Zod allows us to define a validation schema for our form data. This schema is a declarative way of specifying the validation rules for each field. For example, to mark a field as required, we can use the `min(1)` method, which specifies that the field must have a minimum length of 1.
+Zod allows us to define a validation schema for our form data. This schema is a declarative way of specifying the validation rules for each field. For example, to mark a field as required, we can use the `min(1)` method, which specifies that the field must have a minimum length of 1. The 2nd argument is optional and can be used to override the default error message.
+
+Zod has many validation methods for different data types and scenarios, for strings you can use the `email()` method to validate an email address or `url()` to validate a URL, if you have custom needs you can always use `regex()` to validate against a regular expression.
```
Comment (sarmstead) [inline post.md]: Definitely a well-needed sentence, but it feels a little long. Maybe try breaking it up? Something like:

> Zod has many validation methods for different data types and scenarios. You can use the `email()` method to validate an email address, or `url()` to validate a URL. Also, if you have custom needs, you can always use `regex()` to validate against a regular expression.

File: post.md [hunk 12]
```diff
@@ -82,10 +81,11 @@ export default async function contactAction(_prevState: any, params: FormData) {
 
 ## Contact Page
 
-In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a server component while the `Form` component is a client component. We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
+In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a [server component](https://nextjs.org/docs/app/building-your-application/rendering/server-components) while the `Form` component is a [client component](https://nextjs.org/docs/app/building-your-application/rendering/client-components). We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
```
Comment (sarmstead) [inline post.md:84]: Love the links here. Very handy.

File: post.md [hunk 13]
```diff
@@ -163,4 +164,12 @@ const findErrors = (fieldName: string, errors: ZodIssue[]) => {
...
-In conclusion, Zod provides a powerful and flexible way to handle form validation in a Next.js application. By combining it with server-side actions and the `useFormState` hook, we can create robust forms with clear and specific error messages.
+## Wrapping it Up
+
+Zod provides a powerful and flexible way to handle form validation in a Next.js application. We can apply server-side validation to our forms and still present a great user experience by following a few simple steps:
+
+- Define a Zod schema for each server action
+- Validate user input in your server action using the Zod schema
+- Return an array of errors if validation fails
+- In your client component, use the `useFormState` hook to receive validation errors
+- Display the validation errors to the user
```
Comment (sarmstead) [inline post.md]: Much better conclusion, but something still feels a bit off... Maybe try something like this?

> Zod provides a powerful and flexible way to handle form validation in Next.js applications. Using Zod, we can implement server-side validation to our forms and still present a great user experience by following a few simple steps: _[insert steps here]_

File: post.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  "minimum": 1,
+  "type": "string",
+  "inclusive": true,
+  "exact": false,
+  "message": "Name cannot be blank",
+  "path": ["name"]
+}
+```
+
+Now, let's look at how we can implement this in our Next.js application. We'll start with the server action, then move on to the contact page, and finally the form component.
```
Comment (sarmstead) [inline post.md]: Remove comma after _now_. You can even change up the wording:

> Let's look now at how ....

**Pull request merged**